File: raycast-extension/src/ask.ts
import { Action, ActionPanel, Form, Icon, List, showToast, Toast } from "@raycast/api";
import { useState, useEffect } from "react";
import { OhRightResult, OhRightStats, runQuery, runAsk, attentionBar, typeEmoji, checkSetup } from "./shared";

export default function AskCommand() {
  const [searchText, setSearchText] = useState("");
  const [results, setResults] = useState<OhRightResult[]>([]);
  const [stats, setStats] = useState<OhRightStats | null>(null);
  const [message, setMessage] = useState("");
  const [isLoading, setIsLoading] = useState(false);

  const executeQuery = async (query: string) => {
    if (!query.trim()) {
      // Default: show obsessions
      setIsLoading(true);
      const data = runQuery("obsessions");
      setIsLoading(false);
      if (Array.isArray(data)) {
        setResults(data);
        setStats(null);
        setMessage("Your current obsessions:");
      }
      return;
    }

    setIsLoading(true);

    // Quick keyword detection (same as fallback in ask.py for speed)
    const q = query.toLowerCase();
    let command = "search";
    let arg = query;

    if (q.match(/obsess|interest|song|music|listen/)) {
      command = "obsessions";
      arg = "";
      setMessage("Your most played:");
    } else if (q.match(/product|buy|shopping|purchas|forgot/)) {
      command = "products";
      arg = "";
      setMessage("Products you researched:");
    } else if (q.match(/abandon|drop|unfinished|stopped/)) {
      command = "abandoned";
      arg = "";
      setMessage("Things you dropped:");
    } else if (q.match(/stat|graph|summary|overview/)) {
      const data = runQuery("stats");
      setIsLoading(false);
      if (!Array.isArray(data)) {
        setStats(data as OhRightStats);
        setResults([]);
        setMessage("Your behavioral graph:");
      }
      return;
    }

    const data = runQuery(command, arg);
    setIsLoading(false);

    if (Array.isArray(data)) {
      setResults(data);
      setStats(null);
    } else {
      setStats(data as OhRightStats);
      setResults([]);
    }
  };

  // Run on mount
  useEffect(() => {
    executeQuery("");
  }, []);

  return (
    <List
      isLoading={isLoading}
      searchText={searchText}
      onSearchTextChange={(text) => {
        setSearchText(text);
        if (text.length === 0 || text.length > 2) {
          executeQuery(text);
        }
      }}
      searchBarPlaceholder="Ask about your digital life — songs, products, obsessions..."
      throttle
    >
      {message && (
        <List.Section title={message}>
          {results.map((item, i) => (
            <List.Item
              key={i}
              icon={typeEmoji(item.type)}
              title={item.title}
              subtitle={item.subtitle}
              accessories={[
                { text: attentionBar(item.attention || item.match || 0) },
              ]}
              actions={
                <ActionPanel>
                  {item.url && (
                    <Action.OpenInBrowser
                      title="Open in Browser"
                      url={item.url}
                      icon={Icon.Globe}
                    />
                  )}
                  <Action.CopyToClipboard
                    title="Copy Title"
                    content={item.title}
                  />
                  {item.url && (
                    <Action.CopyToClipboard
                      title="Copy URL"
                      content={item.url}
                      icon={Icon.Link}
                    />
                  )}
                </ActionPanel>
              }
              detail={
                <List.Item.Detail
                  metadata={
                    <List.Item.Detail.Metadata>
                      <List.Item.Detail.Metadata.Label
                        title="Type"
                        text={item.type}
                      />
                      <List.Item.Detail.Metadata.Label
                        title="Attention"
                        text={`${((item.attention || 0) * 100).toFixed(0)}%`}
                      />
                      {item.revisits !== undefined && (
                        <List.Item.Detail.Metadata.Label
                          title="Revisits"
                          text={String(item.revisits)}
                        />
                      )}
                      {item.last_seen && (
                        <List.Item.Detail.Metadata.Label
                          title="Last Seen"
                          text={item.last_seen}
                        />
                      )}
                      {item.url && (
                        <List.Item.Detail.Metadata.Link
                          title="Source URL"
                          target={item.url}
                          text={item.url.length > 60 ? item.url.slice(0, 60) + "..." : item.url}
                        />
                      )}
                    </List.Item.Detail.Metadata>
                  }
                />
              }
            />
          ))}
        </List.Section>
      )}

      {stats && (
        <List.Section title={message}>
          <List.Item
            icon={Icon.BarChart}
            title={`${stats.graph.total_nodes} topics tracked`}
            subtitle={`${stats.graph.active} active · ${stats.graph.abandoned} abandoned`}
            accessories={[{ text: `${stats.events.semantic} semantic events` }]}
          />
          {Object.entries(stats.types).map(([type, count]) => (
            <List.Item
              key={type}
              icon={typeEmoji(type)}
              title={type}
              subtitle={`${count} items`}
            />
          ))}
        </List.Section>
      )}

      {!isLoading && results.length === 0 && !stats && searchText && (
        <List.EmptyView
          icon={Icon.MagnifyingGlass}
          title="Nothing found"
          description="Browse more and OhRight will learn your patterns."
        />
      )}

      {!isLoading && results.length === 0 && !stats && !searchText && (
        <List.EmptyView
          icon={Icon.QuestionMark}
          title="OhRight is ready"
          description="Type a query like 'songs I listened to' or 'products I researched'"
        />
      )}
    </List>
  );
}

File: ask.py
#!/usr/bin/env python3
"""
OhRight AI Query Engine
Understands natural language and maps it to graph queries via OpenAI.
Usage: python3 ask.py "fetch me songs I listened to most"
"""

import sys, os, json

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

# Read API key from file or env
def get_openai_key():
    key = os.environ.get("OPENAI_API_KEY", "")
    if not key:
        for path in [
            os.path.expanduser("~/.ohright/.openai_key"),
            os.path.expanduser("~/.ohright/.openai_key"),
        ]:
            if os.path.exists(path):
                with open(path) as f:
                    key = f.read().strip()
                break
    if not key:
        # Try from zshrc
        zshrc = os.path.expanduser("~/.zshrc")
        if os.path.exists(zshrc):
            with open(zshrc) as f:
                for line in f:
                    if "OPENAI_API_KEY" in line:
                        key = line.split("=", 1)[1].strip().strip('"').strip("'")
                        break
    return key


def get_graph_context():
    """Get current graph state as context for the LLM."""
    import subprocess
    r = subprocess.run(
        [sys.executable, os.path.join(os.path.dirname(__file__), "query.py"), "stats"],
        capture_output=True, text=True
    )
    try:
        return json.dumps(json.loads(r.stdout))
    except:
        return "{}"


def understand_query(natural_query: str) -> dict:
    """Use OpenAI to map natural language to a graph query."""
    key = get_openai_key()
    
    # Fallback: basic pattern matching if no key
    if not key:
        return fallback_match(natural_query)
    
    from openai import OpenAI
    client = OpenAI(api_key=key)
    
    context = get_graph_context()
    
    response = client.chat.completions.create(
        model="gpt-4o-mini",
        messages=[{
            "role": "system",
            "content": f"""You are OhRight — a behavioral intelligence system. You map natural language to graph queries.

Available actions and what they do:
- obsessions  → user's current interests, songs, topics, content consumed (most engaged)
- products    → products user researched, purchase intent, things they might buy
- abandoned   → projects/ideas user dropped or stopped working on
- stats       → graph overview and analytics
- recent      → what happened recently
- search      → semantic search with a specific keyword

Current graph state: {context}

Return ONLY valid JSON:
{{"action": "...", "query": "optional search keyword", "message": "friendly one-line intro"}}

Examples:
"songs I listened to most" → {{"action": "obsessions", "message": "Your most played tracks:"}}
"products I forgot to buy" → {{"action": "products", "message": "Products you researched:"}}
"what was I obsessed with" → {{"action": "obsessions", "message": "Your current obsessions:"}}
"show me abandoned projects" → {{"action": "abandoned", "message": "Things you stopped working on:"}}
"my stats" → {{"action": "stats", "message": "Your behavioral graph:"}}
"monitor research" → {{"action": "search", "query": "monitor", "message": "Searching for monitor:"}}"""
        }, {
            "role": "user",
            "content": natural_query
        }],
        temperature=0.1,
        max_tokens=120,
        response_format={"type": "json_object"}
    )
    
    return json.loads(response.choices[0].message.content)


def fallback_match(query: str) -> dict:
    """Basic keyword matching when OpenAI is unavailable."""
    q = query.lower()
    
    if any(w in q for w in ["obsess", "interest", "into", "doing", "song", "music", "listen", "play"]):
        return {"action": "obsessions", "message": "Your current interests:"}
    if any(w in q for w in ["product", "buy", "shopping", "purchas", "missed", "forgot"]):
        return {"action": "products", "message": "Products you researched:"}
    if any(w in q for w in ["abandon", "drop", "unfinished", "stopped"]):
        return {"action": "abandoned", "message": "Projects you dropped:"}
    if any(w in q for w in ["stat", "graph", "summary", "overview"]):
        return {"action": "stats", "message": "Your behavioral graph:"}
    if any(w in q for w in ["recent", "today", "now"]):
        return {"action": "recent", "message": "Recent activity:"}
    
    return {"action": "search", "query": query, "message": f'Searching for "{query}":'}


from typing import Union

def execute_action(action: dict) -> Union[list, dict]:
    """Execute the mapped query against the graph."""
    import subprocess
    
    cmd = action.get("action", "obsessions")
    search_q = action.get("query", "")
    
    if cmd == "search" and search_q:
        r = subprocess.run(
            [sys.executable, os.path.join(os.path.dirname(__file__), "query.py"), "search", search_q],
            capture_output=True, text=True
        )
    else:
        r = subprocess.run(
            [sys.executable, os.path.join(os.path.dirname(__file__), "query.py"), cmd],
            capture_output=True, text=True
        )
    
    try:
        return json.loads(r.stdout)
    except:
        return []


def format_output(data, message: str) -> str:
    """Format results for Raycast display."""
    lines = []
    
    if isinstance(data, list):
        if not data:
            lines.append("Nothing tracked yet. Keep browsing and OhRight will learn.")
        else:
            if message:
                lines.append(message)
                lines.append("")
            for item in data[:8]:
                title = item.get("title", "")[:75]
                sub = item.get("subtitle", "")[:90]
                attn = item.get("attention", 0) or item.get("match", 0)
                bar = "█" * int(attn * 10) + "░" * (10 - int(attn * 10))
                lines.append(f"  {bar}  {title}")
                if sub:
                    lines.append(f"        {sub}")
                # Show URL if available — Raycast auto-detects and makes clickable
                url = item.get("url", "")
                if url:
                    lines.append(f"        🔗 {url}")
                lines.append("")
    elif isinstance(data, dict):
        g = data.get("graph", {})
        e = data.get("events", {})
        t = data.get("types", {})
        if message:
            lines.append(message)
            lines.append("")
        lines.append(f"  Topics tracked:  {g.get('total_nodes', 0)}")
        lines.append(f"  Active:          {g.get('active', 0)}")
        lines.append(f"  Abandoned:       {g.get('abandoned', 0)}")
        lines.append(f"  Events:          {e.get('raw', 0)} raw → {e.get('semantic', 0)} semantic")
        if t:
            top = max(t, key=t.get)
            lines.append(f"  Top category:    {top} ({t[top]})")
        lines.append("")
    else:
        lines.append(str(data))
    
    return "\n".join(lines)


if __name__ == "__main__":
    query = " ".join(sys.argv[1:]) if len(sys.argv) > 1 else "what am I obsessed with"
    
    action = understand_query(query)
    
    data = execute_action(action)
    
    message = action.get("message", "")
    output = format_output(data, message)
    print(output)
